moveit_setup_assistant/src/widgets/setup_assistant_widget.cpp

```diff
@@ -237,118 +227,66 @@ void SetupAssistantWidget::moveToScreen(const int index)
-  connect(controllers_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(controllers_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Simulation Screen
-  simulation_widget_ = new SimulationWidget(this, config_data_);
-  main_content_->addWidget(simulation_widget_);
-  connect(simulation_widget_, SIGNAL(isModal(bool)), this, SLOT(setModalMode(bool)));
-  connect(simulation_widget_, SIGNAL(highlightLink(const std::string&, const QColor&)), this,
-          SLOT(highlightLink(const std::string&, const QColor&)));
-  connect(simulation_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(simulation_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Perception
-  perception_widget_ = new PerceptionWidget(this, config_data_);
-  main_content_->addWidget(perception_widget_);
-  connect(perception_widget_, SIGNAL(isModal(bool)), this, SLOT(setModalMode(bool)));
-  connect(perception_widget_, SIGNAL(highlightLink(const std::string&, const QColor&)), this,
-          SLOT(highlightLink(const std::string&, const QColor&)));
-  connect(perception_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(perception_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Author Information
-  author_information_widget_ = new AuthorInformationWidget(this, config_data_);
-  main_content_->addWidget(author_information_widget_);
-  connect(author_information_widget_, SIGNAL(isModal(bool)), this, SLOT(setModalMode(bool)));
-  connect(author_information_widget_, SIGNAL(highlightLink(const std::string&, const QColor&)), this,
-          SLOT(highlightLink(const std::string&, const QColor&)));
-  connect(author_information_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(author_information_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Configuration Files
-  configuration_files_widget_ = new ConfigurationFilesWidget(this, config_data_);
-  main_content_->addWidget(configuration_files_widget_);
+  if (!plugins_loader_)
+  {
+    try
+    {
+      plugins_loader_.reset(
+          new pluginlib::ClassLoader<SetupScreenWidget>("moveit_setup_assistant", "SetupScreenWidget"));
+    }
+    catch (pluginlib::PluginlibException& ex)
+    {
+      ROS_FATAL_STREAM("Exception while creating octomap updater plugin loader " << ex.what());
+    }
+  }
+
+  // Declared classes with setup screen widget as their base class
+  std::vector<std::string> classes = plugins_loader_->getDeclaredClasses();
+
+  SetupScreenWidget* widget;
+
+  for (size_t i = 0; i < classes.size(); ++i)
+  {
+    try
+    {
+      widget = plugins_loader_->createUnmanagedInstance(classes[i]);
```

Comment:
does createUnmanagedInstance() mean it tracks the memory for you?

Reply:
It doesn't track memory, I really didn't want to use it, but all other plugin instantiation methods failed to work with the QtWidget class, I think because those `instances` must not be destroyed as long as the main QtWidget is running.